lapack/gonum/dlassq.go

```diff
@@ -22,20 +21,90 @@ func (impl Implementation) Dlassq(n int, x []float64, incx int, scale float64, s
 		panic(shortX)
 	}
 
+	if math.IsNaN(scale) || math.IsNaN(sumsq) {
+		return scale, sumsq
+	}
+
+	if sumsq == 0 {
+		scale = 1
+	}
+	if scale == 0 {
+		scale = 1
+		sumsq = 0
+	}
+
 	if n == 0 {
 		return scale, sumsq
 	}
 
-	for ix := 0; ix <= (n-1)*incx; ix += incx {
-		absxi := math.Abs(x[ix])
-		if absxi > 0 || math.IsNaN(absxi) {
-			if scale < absxi {
-				sumsq = 1 + sumsq*(scale/absxi)*(scale/absxi)
-				scale = absxi
-			} else {
-				sumsq += (absxi / scale) * (absxi / scale)
+	// Compute the sum of squares in 3 accumulators:
+	//  - abig: sum of squares scaled down to avoid overflow
+	//  - asml: sum of squares scaled up to avoid underflow
+	//  - amed: sum of squares that do not require scaling
+	// The thresholds and multipliers are:
+	//  - values bigger than dtbig are scaled down by dsbig
+	//  - values smaller than dtsml are scaled up by dssml
+	var (
+		isBig            bool
+		asml, amed, abig float64
+	)
+	for i, ix := 0, 0; i < n; i++ {
+		ax := math.Abs(x[ix])
+		switch {
+		case ax > dtbig:
+			abig += (ax * dsbig) * (ax * dsbig)
```

Comment:
```
ax *= dsbig
abig += ax*ax
```
? (if yes, similar below)

The compiler generates the same code either way.

Reply:
I'm fine with your suggestion. Done at all 4 places.